Automattic/woocommerce-payments
# Fraud Services - Remove hooks from constructor

Fixes #7261 

Currently based on PR #7249 

#### Changes proposed in this Pull Request



Remove hooks from Fraud Services related constructors. Move hooks to new `init_hooks` methods.



This PR moves hooks from the Fraud Services related constructors to new `init_hooks` methods. This PR also updates all calls to call these method. The following classes were updated:

- `Fraud_Risk_Tools`
- `WC_Payments_Fraud_Service`
- `Order_Fraud_And_Risk_Meta_Box`

Calls to instantiate these classes as objects were updated to include calls to `init_hooks` in:

- `WC_Payments`
- `WC_Payments_Fraud_Service_Test`

Lastly, this PR removed the fraud service related exceptions from `dev/phpcs/ruleset.xml`.





#### Testing instructions





* Checkout this branch.
* All tests should pass.
* Add this patch to include some helpful log lines that are written to your client's `debug.log` file while testing these changes: [pr7308.patch](https://github.com/Automattic/woocommerce-payments/files/12742328/pr7308.patch)


#### Testing `Fraud_Risk_Tools`

- The only hook that is moved is the one that initializes the advanced fraud rules page. So verify that you can reach that page on your local: http://localhost:8082/wp-admin/admin.php?page=wc-admin&path=%2Fpayments%2Ffraud-protection.
- You should also see the following entry in the `debug.log` file: Registering the Fraud & Risk Tools Advanced Settings Page.

#### Testing `WC_Payments_Fraud_Service`

- Three hooks' callbacks need to be verified:
  - `prepare_fraud_config`
  - `link_session_if_user_just_logged_in`
  - `add_sift_js_tracker`
- To test the first two, in a browser window login as one of your test customer accounts, add an item to the cart, then go to the checkout page. Then look at your debug log and you should see at least one instance of each of the following two lines (you will see more of the first one as it hooks into `init`):
  - `link_session_if_user_just_logged_in`
  - `prepare_fraud_config`
- To test `add_sift_js_tracker`, simply click around to admin pages that aren't WCPay admin pages. You should see a line logged each time they click to a new page. Also check the elements tab in your browser inspector to see that `WC_Payments_Fraud_Service->add_sift_js_tracker()` is present in a table row for the `print_admin_footer_scripts`.

#### Testing `Order_Fraud_And_Risk_Meta_Box`

- Only one hook is moved to init_hooks: `maybe_add_meta_box`
- To test, click into an order in the admin and you should see the fraud and risk meta box on the right sidebar and the following line should log to `debug.log`: Adding wcpay_order_fraud_and_risk_meta_box.


-------------------


**Post merge**
